In [58]:
!which python

/home/hienpth/miniconda3/envs/p4ds-env/bin/python


In [59]:
#!conda install -y seaborn plotly

# Final project: Phân tích dữ liệu Data Engineer Jobs
Các thành viên:

1. Phạm Thị Hoài Hiền - 18120178

2. Vương Thị Ngọc Linh - 18120195

In [60]:
import numpy as np 
import matplotlib.pyplot as plt
import pandas as pd
import re 
import plotly.express as px

In [61]:
#plt.style.use('seaborn')

## Giới thiệu về dữ liệu

Đây là bộ dữ liệu về việc làm Data Engineer và một số vị trí liên quan, link Kaggle lấy dữ liệu: https://www.kaggle.com/andrewmvd/data-engineer-jobs

Datatset này được tạo bởi Picklesueat và nó chứa hơn 2500 công việc liên quan vị trí Data Engineer.

Dataset gồm có 15 cột:
- Job Tile: Tên vị trí
- Salary Estimate: Mức lương
- Job Description: Miêu tả vị trí
- Rating: Xếp hạng công ty tuyển dụng
- Company name: Tên công ty tuyển dụng
- Location: Địa chỉ nơi làm việc
- Headquarters: Trụ sở của công ty
- Size: Quy mô nhân lực của công ty
- Founded: Năm thành lập công ty
- Type of ownership: Quyền sở hữu
- Industry: Lĩnh vực hoạt động của công ty
- Sector: Lĩnh vực chuyên ngành mà công ty phát triển.
- Revenue: Doanh thu mỗi năm của công ty.
- Competitors: Đối thủ cạnh tranh của công ty.
- Easy Apply: Vị trí này có dễ apply không.


## Tiền xử lí

### Đọc dữ liệu vào dataframe

In [62]:
data_engineer_jobs = pd.read_csv("DataEngineer.csv")
data_engineer_jobs.shape

(2528, 15)

In [63]:
data_engineer_jobs.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Easy Apply'],
      dtype='object')

Dataset có 2528 dòng và 15 cột, đúng như miêu tả.

### Xử lí missing value

Xem dữ liệu, có thể thấy tác giả đã đặt các giá trị thiếu thành `-1` và `Unknown / Non-Applicable` thay vì là `null` hay `nan`.
Đầu tiên chuyển các giá trị đó về thành `nan`.

In [64]:
data_engineer_jobs['Revenue'].unique()

array(['$10 to $25 million (USD)', '$25 to $50 million (USD)',
       '$5 to $10 million (USD)', 'Unknown / Non-Applicable',
       '$100 to $500 million (USD)', '$2 to $5 billion (USD)',
       '$500 million to $1 billion (USD)', '$50 to $100 million (USD)',
       '$10+ billion (USD)', '$1 to $2 billion (USD)',
       '$1 to $5 million (USD)', 'Less than $1 million (USD)',
       '$5 to $10 billion (USD)', '-1'], dtype=object)

In [65]:
data_engineer_jobs.replace({-1: np.nan, '-1': np.nan}, inplace=True)

In [66]:
data_engineer_jobs.isna().sum() / len(data_engineer_jobs) * 100

Job Title             0.000000
Salary Estimate       0.000000
Job Description       0.000000
Rating                9.256329
Company Name          0.000000
Location              0.000000
Headquarters          4.905063
Size                  4.707278
Founded              22.587025
Type of ownership     4.707278
Industry             14.952532
Sector               14.833861
Revenue               4.707278
Competitors          69.185127
Easy Apply           94.224684
dtype: float64

**Xoá các dòng bị thiếu giá trị quá nhiều**

In [67]:
data_engineer_jobs[data_engineer_jobs['Revenue'].isna()]

Job Title  \
114                                Data Engineer   
161                                Data Engineer   
163                                Data Engineer   
165                                Data Engineer   
182                                Data Engineer   
...                                          ...   
2414             Data Engineer - Big Data-Hadoop   
2415           Developer, Data Engineer Big Data   
2419                       Data Network Engineer   
2427                Data Integration Engineer II   
2447  Data Engineer/Irving TX/6+ months Contract   

                   Salary Estimate  \
114    $84K-$113K (Glassdoor est.)   
161   $105K-$121K (Glassdoor est.)   
163   $105K-$121K (Glassdoor est.)   
165   $105K-$121K (Glassdoor est.)   
182   $105K-$121K (Glassdoor est.)   
...                            ...   
2414    $30K-$60K (Glassdoor est.)   
2415    $30K-$60K (Glassdoor est.)   
2419    $30K-$60K (Glassdoor est.)   
2427  $100K-$160K (Glassdoor est.)   
2447  $100K-$160K (Glassdoor est.)   

                                        Job Description  Rating  \
114   We are looking for a data engineer to manage a...     NaN   
161   What You'll Do:We are looking for a data engin...     NaN   
163   Job Details\n\nLevel\n\nExperienced\n\nJob Loc...     NaN   
165   About UsSimon Data was founded in 2015 by a te...     NaN   
182   Role Background:\n\nWere looking for agreat Da...     NaN   
...                                                 ...     ...   
2414  Analyze and understand data sources & APIs• De...     NaN   
2415  OTSI is searching for a Developer, Data Engine...     NaN   
2419  This position will be responsible for the miss...     NaN   
2427  Job Description\nThe Data Integration Engineer...     NaN   
2447  Data Platform Engineer\no Apache Pinot and Apa...     NaN   

                                           Company Name           Location  \
114                              Pierce Technology Corp       New York, NY   
161                                        Atom Finance       New York, NY   
163   Greater New York Mutual Insurance Companies (GNY)       New York, NY   
165                                           Simondata       New York, NY   
182                           Carter Wellington Limited       New York, NY   
...                                                 ...                ...   
2414           Object Technology Solutions, Inc. (OTSI)         Irving, TX   
2415           Object Technology Solutions, Inc. (OTSI)         Irving, TX   
2419                      Brightstar Media Organization  Grand Prairie, TX   
2427                                 Outsourced CIO LLC      Arlington, TX   
2447                            Suncap Technology, Inc.         Irving, TX   

     Headquarters Size  Founded Type of ownership Industry Sector Revenue  \
114           NaN  NaN      NaN               NaN      NaN    NaN     NaN   
161           NaN  NaN      NaN               NaN      NaN    NaN     NaN   
163           NaN  NaN      NaN               NaN      NaN    NaN     NaN   
165           NaN  NaN      NaN               NaN      NaN    NaN     NaN   
182           NaN  NaN      NaN               NaN      NaN    NaN     NaN   
...           ...  ...      ...               ...      ...    ...     ...   
2414          NaN  NaN      NaN               NaN      NaN    NaN     NaN   
2415          NaN  NaN      NaN               NaN      NaN    NaN     NaN   
2419          NaN  NaN      NaN               NaN      NaN    NaN     NaN   
2427          NaN  NaN      NaN               NaN      NaN    NaN     NaN   
2447          NaN  NaN      NaN               NaN      NaN    NaN     NaN   

     Competitors Easy Apply  
114          NaN        NaN  
161          NaN        NaN  
163          NaN        NaN  
165          NaN        NaN  
182          NaN        NaN  
...          ...        ...  
2414         NaN        NaN  
2415         NaN        NaN  
2419         NaN    

Tổng cộng có 119 dòng có giá trị `Revenue` là -1, và có thể thấy các cột còn lại (trừ `Job Title`, `Salary`, `Company Name`, `Location` và `Headquaters`) đểu có giá trị -1, tức là job này và công ty này không có bất cứ thông tin nào có giá trị để đưa vào phân tích. Vì vậy ta sẽ xóa luôn các dòng này.

In [68]:
data_engineer_jobs.drop(data_engineer_jobs[data_engineer_jobs['Revenue'].isna()].index, axis=0, inplace=True)
data_engineer_jobs.shape

(2409, 15)

Dữ liệu còn lại 2409 dòng với cột Revenue không có giá trị -1 mà chỉ còn giá trị Unknown / Non-Applicable, ta sẽ xử lý chúng sau.

**Filling các hàng có cùng tên công ty**

In [69]:
data_engineer_jobs = data_engineer_jobs.sort_values(by=['Company Name','Location','Size','Industry','Sector'])
#data_engineer_jobs.head()

In [70]:
list_of_company = data_engineer_jobs['Company Name'].value_counts(sort=False)
list_of_company

USAA\n3.6                          5
VNE LLC\n3.0                       1
Trabus Technologies                1
ClickUp\n4.6                       1
Red Balloon Security\n4.4          1
                                  ..
InfoQuest Consulting Group\n4.7    1
The Jupiter Group\n4.4             1
Unilever\n4.0                      1
iSphere\n4.8                       2
Staffing the Universe\n2.7         1
Name: Company Name, Length: 1369, dtype: int64

In [71]:
def filling_missing_value(company):
    data_engineer_jobs[data_engineer_jobs['Company Name'] == company] = \
        data_engineer_jobs[data_engineer_jobs['Company Name'] == company].replace(np.nan, value=None, limit=list_of_company[company])
    
    
for company in list_of_company.index:
    filling_missing_value(company)

data_engineer_jobs = data_engineer_jobs.reset_index(drop=True) #do sau khi sort value thì index bị lộn xộn nên cần xếp lại

In [72]:
data_engineer_jobs.isna().sum() / len(data_engineer_jobs) * 100

Job Title             0.000000
Salary Estimate       0.000000
Job Description       0.000000
Rating                4.773765
Company Name          0.000000
Location              0.000000
Headquarters          0.207555
Size                  0.000000
Founded              18.762972
Type of ownership     0.000000
Industry             10.751349
Sector               10.626816
Revenue               0.000000
Competitors          67.662931
Easy Apply           93.939394
dtype: float64

Cột Competitors và Easy Apply có rất nhiều missing value, và dữ liệu hai cột này cũng không cần thiết cho câu hỏi, vì vậy sẽ xóa hai cột này để làm sạch dữ liệu.

In [73]:
data_engineer_jobs = data_engineer_jobs.drop(columns=['Competitors', 'Easy Apply'])

### Xử lí các cột numeric

In [74]:
data_engineer_jobs.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue'],
      dtype='object')

Có đầy đủ các cột và nó có tên đúng như mô tả. Tiếp theo ta sẽ xem kiểu dữ liệu của cột.

In [75]:
data_engineer_jobs.dtypes

Job Title             object
Salary Estimate       object
Job Description       object
Rating               float64
Company Name          object
Location              object
Headquarters          object
Size                  object
Founded              float64
Type of ownership     object
Industry              object
Sector                object
Revenue               object
dtype: object

Theo lí thuyết ta sẽ phải có những cột này là kiểu numeric: `Salary Estimate`, `Rating`, `Founded`, `Size`, `Revenue`. Tuy nhiên theo kết quả trên thì các cột `Salary Estimate`, `Size` và `Revenue` chưa có đúng kiểu dữ liệu (object thay vì int hoặc float). Ta tiến hành đổi kiểu dữ liệu lại cho 3 cột này.

**Đổi kiểu dữ liệu cột `Salary Estimate`**

In [76]:
data_engineer_jobs["Salary Estimate"].unique()

array(['$54K-$104K (Glassdoor est.)', '$79K-$104K (Glassdoor est.)',
       '$28K-$53K (Glassdoor est.)', '$67K-$107K (Glassdoor est.)',
       '$59K-$135K (Glassdoor est.)', '$67K-$72K (Glassdoor est.)',
       '$45K-$76K (Glassdoor est.)', '$48K-$74K (Glassdoor est.)',
       '$80K-$150K (Glassdoor est.)', '$71K-$130K(Employer est.)',
       '$74K-$123K (Glassdoor est.)', '$52K-$63K (Glassdoor est.)',
       '$57K-$108K (Glassdoor est.)', '$94K-$166K (Glassdoor est.)',
       '$54K-$113K (Glassdoor est.)', '$55K-$105K (Glassdoor est.)',
       '$39K-$77K (Glassdoor est.)', '$67K-$126K (Glassdoor est.)',
       '$30K-$60K (Glassdoor est.)', '$27K-$50K (Glassdoor est.)',
       '$69K-$122K (Glassdoor est.)', '$77K-$181K (Glassdoor est.)',
       '$113K-$198K (Glassdoor est.)', '$46K-$92K (Glassdoor est.)',
       '$106K-$127K (Glassdoor est.)', '$140K-$151K (Glassdoor est.)',
       '$94K-$107K (Glassdoor est.)', '$43K-$84K (Glassdoor est.)',
       '$73K-$94K (Glassdoor est.)', '$65K-

Tất cả các giá trị đều được biểu diễn trong một khoảng cố định. `Glassdoor est.` là mức lương ước tính trên trang việc làm trực tuyến `Glassdoor`. 
Ta đổi giá trị của cột `Salary Estimate` từ mức lương min-max thành mức lương trung bình để dễ dàng so sánh và tính toán.
Rất may là cột này không có giá trị thiếu.

In [77]:
def find_mean_salary(x):
    S = [int(s) for s in re.findall(r'\d+', x)] #liệt kê (list) tất cả số nguyên trong x
    return (S[0] + S[1]) / 2

data_engineer_jobs["Salary Estimate"] = data_engineer_jobs["Salary Estimate"].apply(find_mean_salary)
data_engineer_jobs = data_engineer_jobs.rename(columns={"Salary Estimate": "Salary est. (thousand USD)"})

**Đổi kiểu dữ liệu cột `Size`**

In [78]:
data_engineer_jobs['Size'].unique()

array(['1 to 50 employees', '51 to 200 employees',
       '1001 to 5000 employees', '201 to 500 employees',
       '10000+ employees', '5001 to 10000 employees',
       '501 to 1000 employees', 'Unknown'], dtype=object)

Ta thấy chỉ có 7 giá trị phân biệt và 1 giá trị missing (unknown). 
Vậy ta quy đổi thành các giá trị từ 1-7 thể hiện cho quy mô tăng dần như sau:
- Mức độ 1: 1 to 50 employees
- Mức độ 2: 51 to 200 employees
- Mức độ 3: 201 to 500 employees
- Mức độ 4: 501 to 1000 employees
- Mức độ 5: 1001 to 5000 employees
- Mức độ 6: 5001 to 10000 employees
- Mức độ 7: 10000+ employees
Các giá trị unknown ta sẽ để bằng `nan`.

In [79]:
def transform_size(x):
    if x == '1 to 50 employees':
        return 1
    elif x == '51 to 200 employees':
        return 2
    elif x == '201 to 500 employees':
        return 2
    elif x == '501 to 1000 employees':
        return 4
    elif x == '1001 to 5000 employees':
        return 5
    elif x == '5001 to 10000 employees':
        return 6
    elif x == '10000+ employees':
        return 7
    else:
        return np.nan

data_engineer_jobs['Size'] = data_engineer_jobs['Size'].apply(transform_size)

**Đổi kiểu dữ liệu cột `Revenue`**

In [80]:
data_engineer_jobs['Revenue'].unique()

array(['Unknown / Non-Applicable', '$5 to $10 million (USD)',
       '$25 to $50 million (USD)', 'Less than $1 million (USD)',
       '$1 to $2 billion (USD)', '$500 million to $1 billion (USD)',
       '$10+ billion (USD)', '$50 to $100 million (USD)',
       '$100 to $500 million (USD)', '$5 to $10 billion (USD)',
       '$2 to $5 billion (USD)', '$10 to $25 million (USD)',
       '$1 to $5 million (USD)'], dtype=object)

Ta thấy có rất nhiều cách biểu diễn khác nhau cho cột này (có khoảng from..to, có con số X+, có less than). 
Do số tiền chỉ có đơn vị là triệu (million) và tỷ (billion) nên ta sẽ đổi hết giá trị về thành đơn vị tỉ đô
Đổi giá trị cột `Revenue` từ các khoảng doanh thu bằng `giá trị cao nhất (max)` trong khoảng:
- Các dòng có giá trị X+ sẽ được quy ước là X+1
- Các dòng có giá trị Less than X quy ước là X 
- Missing value sẽ được quy ước là 0

In [81]:
def find_max_revenue(x):
    if x == 'Unknown / Non-Applicable':
        return -1
    else:
        S = [int(s) for s in re.findall(r'\d+', x)]
        if 'billion' in x and 'million' not in x:
            S = [s*1000 for s in S]
        if '+' in x:
            S.append(S[0] + 1)
        if 'Less than' in x:
            S.append(S[0])
            S[0] = 0
        return S[1]
data_engineer_jobs['Revenue'] = data_engineer_jobs['Revenue'].apply(find_max_revenue)
data_engineer_jobs = data_engineer_jobs.rename(columns={'Revenue': 'Revenue (million USD)'})

### Xử lí các cột category

Ở cột `Company Name`, ta thấy có phần "\n rating" phía sau tên công ty là dư thừa (vì đã có cột `rating` riêng), vì vậy ta xoá nó đi.

In [82]:
data_engineer_jobs['Company Name'] = data_engineer_jobs['Company Name'].str.partition("\n")

Ta xem xét một vài thống kê trong các cột category:

In [83]:
data_engineer_jobs.describe(include=[object])  

Job Title                                    Job Description  \
count            2409                                               2409   
unique           1221                                               2190   
top     Data Engineer  Organization: Accenture Federal Services\n\nLo...   
freq              438                                                  7   

       Company Name    Location  Headquarters  Type of ownership     Industry  \
count          2409        2409          2404               2409         2150   
unique         1365         111           435                 13           83   
top          Amazon  Austin, TX  New York, NY  Company - Private  IT Services   
freq             28         195           178               1392          355   

                        Sector  
count                     2153  
unique                      24  
top     Information Technology  
freq                       890

### Xử lí cột `Job Title`

Ở đây ta quan tâm đến 2 job là Data Engineer và Senior Data Engineer, và có rất nhiều job thuộc 2 loại này nhưng có title khác nhau, ví dụ: Big Data Engineer, Data Engineer II,... Vì vậy cần đồng bộ các title này trở về 2 loại chính. Đầu tiên là sửa lại title chứa Sr. thành Senior để dễ xử lý.

In [84]:
data_engineer_jobs['Job Title'] = data_engineer_jobs['Job Title'].str.replace('Sr.','Senior', regex=True)

Tiếp theo ta thay đổi tất cả title chứa cụm 'Data Engineer' -> Data Engineer và tương tự với Senior Data Engineer và lưu lại vào cột Job Title.

In [85]:
title = []
job_title = data_engineer_jobs['Job Title'].values
for each in job_title:
    if 'Senior Data Engineer' not in each:
        #lệnh re replace các title Data Engineer
        x = re.sub("[\w\s\S]*Data[\w\s\S]*Engineer[\w\s\S]*",'Data Engineer',each,flags=re.IGNORECASE) 
    else:
        #lệnh re thay đổi các title Senior Data Engineer
        x = re.sub("[\w\s\S]*Senior[\w\s\S]*Data[\w\s\S]*Engineer[\w\s\S]*",'Senior Data Engineer', each,flags=re.IGNORECASE)
    title.append(x)

In [86]:
data_engineer_jobs['Job Title']= title
data_engineer_jobs['Job Title'].value_counts()

Data Engineer                                   1191
Senior Data Engineer                             146
Software Engineer                                 90
Machine Learning Engineer                         17
Systems Engineer                                  13
                                                ... 
Senior Software Engineer $5000 SIGN ON BONUS       1
Senior Thermal Modeling Engineer                   1
Matlab Software Engineer                           1
SeniorInformation Security Engineer                1
Software Systems Engineer I                        1
Name: Job Title, Length: 745, dtype: int64

## Các câu hỏi cần trả lời


### 1- Mức lương trung bình của từng job là bao nhiêu? Mức lương cho Senior data là bao nhiêu?

In [87]:
temp = data_engineer_jobs[["Job Title","Salary est. (thousand USD)"]].groupby(["Job Title"]).mean().sort_values(by=['Salary est. (thousand USD)'], ascending=False)
temp

Salary est. (thousand USD)
Job Title                                                                     
machine learning Engineer with AWS                                       217.5
Python Software Engineer                                                 217.5
Software Engineer III or IV                                              204.0
Data Analyst - EDSS                                                      204.0
Software Engineer - Developer Experience                                 204.0
...                                                                        ...
Equipment Engineer                                                        38.5
Engineer Quality Control Inspector with Securit...                        38.5
Software Engineer/Software Developer                                      38.5
Software Support Engineer - C#                                            38.5
Process Optimization Engineer Assistant Vice Pr...                        38.5

[745 rows x 1 columns]

In [88]:
temp.head(50).plot.bar(figsize=(15,3))

<AxesSubplot:xlabel='Job Title'>

Vì có đến hơn 700 job khác nhau nên chỉ show mức lương trung bình của top 50 job đầu (cũng là 50 job được tuyển nhiều nhất).

In [89]:
temp[temp.index == 'Senior Data Engineer'].values

array([[108.00684932]])

Như vậy, Senior Data Engineer có mức lương trung bình là 108 ngàn USD/năm

In [90]:
temp.index

Index(['machine learning Engineer with AWS', 'Python Software Engineer',
       'Software Engineer III or IV', 'Data Analyst - EDSS',
       'Software Engineer - Developer Experience',
       'Full-time Software Engineer',
       'Software Architect (Distinguished Software Engineer) Data Integration',
       'Software Architect (Distinguished Software Engineer) - Data Integration',
       'Cell Culture Bioprocess Development Engineer',
       'Senior Staff Engineer, CAD',
       ...
       'Java Software Engineer III', 'Warehouse IT Support Engineer',
       'Materials Engineer', 'SeniorIT Systems Engineer',
       'Junior Site Reliability Engineer', 'Equipment Engineer',
       'Engineer Quality Control Inspector with Security Clearance',
       'Software Engineer/Software Developer',
       'Software Support Engineer - C#',
       'Process Optimization Engineer Assistant Vice President'],
      dtype='object', name='Job Title', length=745)

### 2 - Top 5 lĩnh vực tuyển dụng data engineer (tất cả các vị trí) nhiều nhất?

In [91]:
industry = data_engineer_jobs[data_engineer_jobs['Job Title'] == "Data Engineer"]["Industry"].value_counts()
#industry.plot.bar(figsize=(20,8))
industry

IT Services                                236
Staffing & Outsourcing                     110
Computer Hardware & Software                91
Internet                                    73
Enterprise Software & Network Solutions     68
                                          ... 
Other Retail Stores                          1
Trucking                                     1
Chemical Manufacturing                       1
Construction                                 1
Oil & Gas Exploration & Production           1
Name: Industry, Length: 66, dtype: int64

**5 lĩnh vực tuyển dụng data engineer nhiều nhất là:**

In [92]:
industry.head(5)

IT Services                                236
Staffing & Outsourcing                     110
Computer Hardware & Software                91
Internet                                    73
Enterprise Software & Network Solutions     68
Name: Industry, dtype: int64

### 3 - Top 10 location tuyển dụng nhiều vị trí nhất?

In [93]:
#Code
location = data_engineer_jobs['Location'].value_counts()
location.head(10)

Austin, TX          195
Houston, TX         190
San Antonio, TX     185
San Diego, CA       184
Chicago, IL         181
New York, NY        181
Philadelphia, PA    128
Jacksonville, FL    112
Phoenix, AZ         110
Fort Worth, TX      101
Name: Location, dtype: int64

### 4 - Lĩnh vực nào có mức lương trung bình (trên tất cả job) cao nhất và thấp nhất?

In [94]:
#Code

### 5 - Các skill cần thiết cho các job là gì?

In [95]:
pg_req = ["python", "c++", "java", "matlab", ".net", "c#", "javascript", "html", "bash"]
pg_lan = np.zeros(len(pg_req))

In [96]:
def count_pg_lan(language):
    if language == 'c++':
        language = 'c\+\+'
    elif language == '.net':
        language = '\.net'
    elif language == 'c#':
        language = 'c\#'
    tmp = data_engineer_jobs['Job Description'].str.lower()
    return tmp.str.count(language, flags=re.IGNORECASE).sum()

for req in pg_req:
    pg_lan[pg_req.index(req)] = count_pg_lan(req)

pg_lan_df = pd.DataFrame([pg_req, pg_lan]).T
pg_lan_df = pg_lan_df.rename(columns={0: 'Programming Language', 1: 'Count'})
pg_lan_df.sort_values(['Count'], axis=0, ascending=False, inplace=True)
pg_lan_df.set_index('Programming Language', inplace=True)
pg_lan_df



Count
Programming Language        
python                1940.0
java                  1457.0
c++                    302.0
javascript             300.0
.net                   190.0
c#                     171.0
html                   149.0
bash                   127.0
matlab                  55.0

In [97]:
pg_lan_df.plot.bar(figsize=(12,7))

<AxesSubplot:xlabel='Programming Language'>

In [98]:
ds_ml_req = ["machine learning","regression","stat","numpy","pandas","data visualization","data analysis"," ai ","nlp","deep learning"]
ds_ml_count = np.zeros(len(ds_ml_req))
def count_skills_ds_ml(skill):
    tmp = data_engineer_jobs['Job Description'].str.lower()
    if skill == 'machine learning':
        return tmp.str.count(skill, flags=re.IGNORECASE).sum() + tmp.str.count(' ml ', flags=re.IGNORECASE).sum()
    elif skill == 'deep learning':
        return tmp.str.count(skill, flags=re.IGNORECASE).sum() + tmp.str.count(' dl ', flags=re.IGNORECASE).sum()
    return tmp.str.count(skill, flags=re.IGNORECASE).sum()

for req in ds_ml_req:
    ds_ml_count[ds_ml_req.index(req)] = count_skills_ds_ml(req)

ds_ml_df = pd.DataFrame([ds_ml_req, ds_ml_count]).T
ds_ml_df = ds_ml_df.rename(columns={0: 'DS/ML skill', 1: 'Count'})
ds_ml_df.sort_values(['Count'], axis=0, ascending=False, inplace=True)
ds_ml_df.set_index('DS/ML skill', inplace=True)
ds_ml_df

Count
DS/ML skill               
stat                3465.0
machine learning     997.0
data analysis        298.0
data visualization   191.0
 ai                  145.0
deep learning        112.0
regression            84.0
nlp                   60.0
pandas                55.0
numpy                 38.0

In [99]:
ds_ml_df.plot.bar(figsize=(12,7))

<AxesSubplot:xlabel='DS/ML skill'>

In [100]:
edu_req = ["master","doctorate","bachelor"]
edu_count = np.zeros(len(edu_req))
def count_edu(degree):
    tmp = data_engineer_jobs['Job Description'].str.lower()
    if degree == 'master':
        return tmp.str.count(degree, flags=re.IGNORECASE).sum() + tmp.str.count(' ms ', flags=re.IGNORECASE).sum() \
                + tmp.str.count('m.s', flags=re.IGNORECASE).sum()
    elif degree == 'bachelor':
        return tmp.str.count(degree, flags=re.IGNORECASE).sum() + tmp.str.count(' bs ', flags=re.IGNORECASE).sum() \
                + tmp.str.count('b.s', flags=re.IGNORECASE).sum()
    else:
        return tmp.str.count(degree, flags=re.IGNORECASE).sum() + tmp.str.count('phd', flags=re.IGNORECASE).sum() \
                + tmp.str.count('ph.d', flags=re.IGNORECASE).sum()


for req in edu_req:
    edu_count[edu_req.index(req)] = count_edu(req)

edu_df = pd.DataFrame([edu_req, edu_count]).T
edu_df = edu_df.rename(columns={0: 'Education', 1: 'Count'})
edu_df.sort_values(['Count'], axis=0, ascending=False, inplace=True)
edu_df.set_index('Education', inplace=True)
edu_df

Count
Education         
bachelor   14144.0
master      8496.0
doctorate    180.0

In [101]:
edu_df.plot.bar(figsize=(10,5))

<AxesSubplot:xlabel='Education'>

In [102]:
other_skills_req = [" sql ", "nosql", "algorithm", "excel", "sas", "tableau", "power bi", "communication"]
other_skills_count = np.zeros(len(other_skills_req))
def count_other_skills(skill):
    tmp = data_engineer_jobs['Job Description'].str.lower()
    if skill == 'power bi':
        return tmp.str.count(skill, flags=re.IGNORECASE).sum() + tmp.str.count('powerbi', flags=re.IGNORECASE).sum()
    else:
        return tmp.str.count(skill, flags=re.IGNORECASE).sum()

for req in other_skills_req:
    other_skills_count[other_skills_req.index(req)] = count_other_skills(req)

other_skills_df = pd.DataFrame([other_skills_req, other_skills_count]).T
other_skills_df = other_skills_df.rename(columns={0: 'Other skills', 1: 'Count'})
other_skills_df.sort_values(['Count'], axis=0, ascending=False, inplace=True)
other_skills_df.set_index('Other skills', inplace=True)
other_skills_df


Count
Other skills         
communication  1397.0
 sql           1258.0
excel          1179.0
algorithm       512.0
nosql           455.0
tableau         354.0
sas             271.0
power bi        144.0

In [103]:
other_skills_df.plot.bar(figsize=(12,7))

<AxesSubplot:xlabel='Other skills'>

### 6 - Mối quan hệ giữa rating và mức lương trung bình?

In [104]:
fig = px.scatter(data_engineer_jobs, x='Rating', y='Salary est. (thousand USD)',color="Rating",\
                hover_data=['Headquarters', 'Type of ownership', 'Sector','Company Name'], \
                title = "Data Engineer jobs",\
                width=800, height=500)
fig.show()